storagenode/storagenodedb/pieceexpiration.go

```diff
@@ -97,3 +98,28 @@ func (db *pieceExpirationDB) DeleteFailed(ctx context.Context, satelliteID storj
 	`, when.UTC(), satelliteID, pieceID)
 	return ErrPieceExpiration.Wrap(err)
 }
+
+// Trash marks a piece expiration as "trashed"
+func (db *pieceExpirationDB) Trash(ctx context.Context, satelliteID storj.NodeID, pieceID storj.PieceID) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	_, err = db.ExecContext(ctx, `
+		UPDATE piece_expirations
+			SET trash = 1
+			WHERE satellite_id = ?
+				AND piece_id = ?
+	`, satelliteID, pieceID)
+	return ErrPieceExpiration.Wrap(err)
+}
+
+// Restore restores all trashed pieces
+func (db *pieceExpirationDB) RestoreTrash(ctx context.Context, satelliteID storj.NodeID) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	_, err = db.ExecContext(ctx, `
+		UPDATE piece_expirations
+			SET trash = 0
+			WHERE satellite_id = ?
+	`, satelliteID)
```

Comment:
This query will be expensive because it will require walking all of the pieces for a given satellite. I suggest we use a [partial index](https://www.sqlite.org/partialindex.html). Specifically,

```
CREATE INDEX piece_expirations_trashed
    ON piece_expirations(satellite_id, trash)
    WHERE trash = 1
```

then this query would become

```
UPDATE piece_expirations
    SET trash = 0
    WHERE satellite_id = ?
    AND trash = 1
```

Then we don't have bloating of indexes for pieces that aren't trash, and fast queries for if we need to restore.

Reply:
Does the index help much, given that the cardinality of this field is 2?

Reply:
Wait, now I'm confused. If there's not an entry for every single piece, then the code to mark it as trash possibly has to do an insert, right?

Reply:
Now I'm doubly confused. So the cardinality of importance is not of the `trash` column, it's how many rows you have to look at to set `trash` to 0. If you have 1,000,000 rows for some satellite that are not trash, and 1 row that is trash, this UPDATE will look at 1,000,001 rows without an index.

Reply:
That makes sense, I thought it was only the cardinality of the `trash` column that matters. I'll add an index.

Reply:
Well, also, I don't think we have an entry for every piece in the piece_expirations table, only ones that actually have an expiration. This conversation made me realize that.

Reply:
That is correct. Not every piece has an expiration, but we don't do an answer to mark as trash, because we only need to mark the _expiration entry_ as trash if it exists.

Reply:
Oh no why are we adding more things to the database? Can't we use file mtime? Sorry I'll read through the PR.

Reply:
@thepaul let me know your thoughts after looking through the PR. I need a way to "trash" the expiration entry, but also to restore it later if a restore call comes in. That's why I added "more things" to the database.

I'm not sure what you mean by using file mtime, can you explain more? I don't see how that helps with the expiration entry.

Reply:
Sorry, don't worry about that, I was misunderstanding the snippet and thought we were adding a new table for tracking trash.

Still would like to review this thoroughly but it will take me some time. I know you want to get it merged but I'm gun-shy about rushing something that adds this sort of complexity. :(

Reply:
I'm not in a rush on this one. The one that's been active for a long time is https://github.com/storj/storj/pull/3529. That one I would indeed like to get merged, but this one was posted today and I always want thorough reviews. :)

Reply:
Oh! Okay. Cool

Reply:
@zeebo I added that index.